### PR TITLE
[ADD] formulas: introduce `TEXTSPLIT` function

### PR DESCRIPTION
**Pull request by Rachico**
The current `SPLIT` function (inspired by Google Sheets) allows splitting a text using delimiters, but it does not exist in Excel. This causes errors when importing sheets containing the `SPLIT` function into Excel. This commit introduces the `TEXTSPLIT`function to enrich the library and ensure better compatibility with Excel.

Task: [4766568](https://www.odoo.com/odoo/2328/tasks/4766568)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_t("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

### CODE REVIEW

File: src/functions/module_text.ts [hunk 1]
```diff
@@ -436,6 +436,112 @@ export const SPLIT = {
   isExported: true,
 } satisfies AddFunctionDescription;
 
+// -----------------------------------------------------------------------------
+// TEXTSPLIT
+// -----------------------------------------------------------------------------
+
+const TEXTSPLIT_DEFAULT_IGNORE_EMPTY = false;
+const TEXTSPLIT_DEFAULT_MATCH_MODE = 0;
+const TEXTSPLIT_DEFAULT_PAD_WITH = "N/A";
```
Comment (hokolomopo) [inline src/functions/module_text.ts:445]:
```suggestion
const TEXTSPLIT_DEFAULT_PAD_WITH = "#N/A";
```

Comment (hokolomopo) [reply src/functions/module_text.ts:445]: And we should return actual #N/A errors, not cells with  #N/A string

File: src/functions/module_text.ts [hunk 2]
```diff
@@ -436,6 +436,112 @@ export const SPLIT = {
...
+
+export const TEXTSPLIT = {
+  description: _t("Splits text into rows or columns using specified column and row delimiters."),
+  args: [
+    arg("text (string)", _t("The text to split.")),
+    arg(
+      "col_delimiter (string, range<string>, optional)",
```
Comment (hokolomopo) [inline src/functions/module_text.ts:452]: arg is marked as mandatory in excel

File: src/functions/module_text.ts [hunk 3]
```diff
@@ -436,6 +436,112 @@ export const SPLIT = {
...
+// -----------------------------------------------------------------------------
+
+const TEXTSPLIT_DEFAULT_IGNORE_EMPTY = false;
+const TEXTSPLIT_DEFAULT_MATCH_MODE = 0;
+const TEXTSPLIT_DEFAULT_PAD_WITH = "N/A";
+
+export const TEXTSPLIT = {
+  description: _t("Splits text into rows or columns using specified column and row delimiters."),
+  args: [
+    arg("text (string)", _t("The text to split.")),
```
Comment (hokolomopo) [inline src/functions/module_text.ts:450]: excel return #VALUE if the text is empty. we return an empty string

File: src/functions/module_text.ts [hunk 4]
```diff
@@ -436,6 +436,112 @@ export const SPLIT = {
...
+    colDelimiter: Arg,
+    rowDelimiter: Arg,
+    ignoreEmpty: Maybe<FunctionResultObject> = { value: TEXTSPLIT_DEFAULT_IGNORE_EMPTY },
+    matchMode: Maybe<FunctionResultObject> = { value: TEXTSPLIT_DEFAULT_MATCH_MODE },
+    padWith: Maybe<FunctionResultObject> = { value: TEXTSPLIT_DEFAULT_PAD_WITH }
+  ) {
+    const _text = toString(text);
+    const _colDelimiter = toMatrix(colDelimiter)
+      .flat()
+      .map((v) => escapeRegExp(toString(v)));
```
Comment (hokolomopo) [inline src/functions/module_text.ts:483]:
```suggestion
      .flatMap((v) => escapeRegExp(toString(v)));
```

Comment (hokolomopo) [reply src/functions/module_text.ts:483]: same for rows

File: src/functions/module_text.ts [hunk 5]
```diff
@@ -436,6 +436,112 @@ export const SPLIT = {
...
+    const splitByEachColDelimiter = _colDelimiter.length > 1;
+    const _rowDelimiter = toMatrix(rowDelimiter)
+      .flat()
+      .map((v) => escapeRegExp(toString(v)));
+    const splitByEachRowDelimiter = _rowDelimiter.length > 1;
```
Comment (hokolomopo) [inline src/functions/module_text.ts:488]: nitpick: could inline the condition where we need it. I don't think the separate variable helps the comprehension

File: src/functions/module_text.ts [hunk 6]
```diff
@@ -436,6 +436,112 @@ export const SPLIT = {
...
+        ? new RegExp(_colDelimiter.join("|"), regexpFlags)
+        : new RegExp(_colDelimiter[0], regexpFlags);
+      const columns = _text.split(colRegexp);
+      cells[0] = [];
+      for (const col of columns) {
+        if (col === "") {
+          if (_ignoreEmpty) {
+            continue;
+          }
+        }
```
Comment (hokolomopo) [inline src/functions/module_text.ts:530]:
```suggestion
        if (col === "" && _ignoreEmpty) {
            continue;
          
        }
```

File: src/functions/module_text.ts [hunk 7]
```diff
@@ -436,6 +436,112 @@ export const SPLIT = {
...
+    }
+
+    let cells: string[][] = [];
+    const regexpFlags = _matchMode === 1 ? "gi" : "g";
+
+    if (_rowDelimiter.length > 0) {
+      const rowRegexp = splitByEachRowDelimiter
+        ? new RegExp(_rowDelimiter.join("|"), regexpFlags)
+        : new RegExp(_rowDelimiter[0], regexpFlags);
+      const rowParts = _text.split(rowRegexp);
```
Comment (hokolomopo) [inline src/functions/module_text.ts:504]: you're not applying _ignoreEmpty if there is a row delimiter but no col delimiter

File: src/functions/module_text.ts [hunk 8]
```diff
@@ -436,6 +436,112 @@ export const SPLIT = {
...
+      if (_colDelimiter.length > 0) {
+        for (const rowText of rowParts) {
+          const colRegexp = splitByEachColDelimiter
+            ? new RegExp(_colDelimiter.join("|"), regexpFlags)
+            : new RegExp(_colDelimiter[0], regexpFlags);
```
Comment (hokolomopo) [inline src/functions/module_text.ts:509]: no need to create a new regex at each loop iteration is there ?

File: src/functions/module_text.ts (hunk 6 above)
Comment (hokolomopo) [reply src/functions/module_text.ts:530]: or use `.filter` ?

File: src/functions/module_text.ts (hunk 7 above)
Comment (hokolomopo) [reply src/functions/module_text.ts:504]: and the _ignoreEmpty is wrong if there is both col & row delimiter, I can end up with a row full of #N/A which should not happens

File: src/functions/module_text.ts [hunk 9]
```diff
@@ -436,6 +436,112 @@ export const SPLIT = {
...
+        cells[0].push(col);
+      }
+    }
+    const maxLength = Math.max(...cells.map((row) => row.length));
+    for (const row of cells) {
+      while (row.length < maxLength) {
+        row.push(toString(_padWith));
+      }
+    }
+    return transposeMatrix(cells);
```
Comment (hokolomopo) [inline src/functions/module_text.ts:540]: nitpick: performance-wise, is there a way to apply the split in a different order so we don't have to transpose at the end? If it's too annoying/make the code more complex, it's fine to keep the transpose

File: tests/functions/module_text.test.ts [hunk 10]
```diff
@@ -571,6 +571,94 @@ describe("SPLIT function", () => {
...
+    const model = createModelFromGrid(grid);
+    setCellContent(model, "F1", ", ");
+    setCellContent(model, "F2", ";");
+    setCellContent(model, "A2", "=TEXTSPLIT(A1, F1:F2)");
+    expect(getRangeValuesAsMatrix(model, "A2:E2")).toEqual([
+      ["Apple", "Banana", "Orange", "Grapes", "Tomato"],
+    ]);
+  });
+
+  test("Ignore empty values", () => {
```
Comment (hokolomopo) [inline tests/functions/module_text.test.ts:629]: missing test for "Ignore empty values" with row delimiter, and row & col delimiter since they do not work :)

File: tests/functions/module_text.test.ts [hunk 11]
```diff
@@ -571,6 +571,94 @@ describe("SPLIT function", () => {
...
+    const grid = { A1: "Dog,,Cat,,Bird" };
+    const model = createModelFromGrid(grid);
+    setCellContent(model, "A2", '=TEXTSPLIT(A1, ",", , TRUE)');
+    expect(getRangeValuesAsMatrix(model, "A2:C2")).toEqual([["Dog", "Cat", "Bird"]]);
+  });
+
+  test("Case-insensitive split", () => {
+    const grid = { A1: "AppleDELIMbanaNADelimcHerry" };
+    const model = createModelFromGrid(grid);
+    setCellContent(model, "A2", '=TEXTSPLIT(A1, "delim",,,1)');
```
Comment (hokolomopo) [inline tests/functions/module_text.test.ts:639]: also test what happens if the argument is set to 0

File: tests/functions/module_text.test.ts [hunk 12]
```diff
@@ -571,6 +571,94 @@ describe("SPLIT function", () => {
...
+      ["Name", "Age"],
+      ["City", "Country"],
+    ]);
+  });
+
+  test("Multiple delimiters", () => {
+    const grid = { A1: "Apple, Banana;Orange, Grapes;Tomato" };
+    const model = createModelFromGrid(grid);
+    setCellContent(model, "F1", ", ");
+    setCellContent(model, "F2", ";");
```
Comment (hokolomopo) [inline tests/functions/module_text.test.ts:622]: nitpick: you could set those in the grid just above

**Review by hokolomopo — COMMENTED** [review]
👋 

ignore_empty parameters seems a bit bugged unfortunately :x

File: src/functions/module_text.ts (hunk 5 above)
Comment (Rachico) [reply src/functions/module_text.ts:488]: Yes that's what I did in the beginning but I think it's more readable this way :)

File: src/functions/module_text.ts [hunk 13]
```diff
@@ -436,6 +436,120 @@ export const SPLIT = {
...
+    if (_rowDelimiter.length > 0) {
+      const rowRegexp = splitByEachRowDelimiter
+        ? new RegExp(_rowDelimiter.join("|"), regexpFlags)
+        : new RegExp(_rowDelimiter[0], regexpFlags);
+      let rowParts = _text.split(rowRegexp);
+      if (_ignoreEmpty) {
+        rowParts = rowParts.filter((v) => v !== "");
+      }
+      if (_colDelimiter.length > 0) {
+        const colRegexp = splitByEachColDelimiter
```
Comment (laa-odoo) [inline src/functions/module_text.ts:514]: does `const colRegexp =  new RegExp(_colDelimiter.join("|"), regexpFlags)` naturally handles the default case where `_colDelimiter.length` is equal to 1 ? same with `splitByEachRowDelimiter` ?

in this case, we can remove `splitByEachRowDelimiter` and `splitByEachColDelimiter`

**Review by rrahir — COMMENTED** [review]
just some fine tuning for the error messages, otherwise good 👍

File: src/functions/module_text.ts [hunk 14]
```diff
@@ -526,6 +526,105 @@ export const TEXTJOIN = {
...
+    }
+
+    const maxLength = Math.max(...cells.map((row) => row.length));
+    for (const row of cells) {
+      while (row.length < maxLength) {
+        row.push(padWith);
+      }
+    }
+
+    return transposeMatrix(cells);
```
[the remark this answers is not in the document]
Comment (laa-odoo) [reply src/functions/module_text.ts:623]: I'm not a fan either, but it's already used a large number of times in all the formulas. A performance improvement task is planned where we need to reduce, if possible, calls to `transposeMatrix` and, more generally, to `generateMatrix`, which is behind it.

Here, we've already significantly reduced the size of the compute to make it easier to understand (because there are many subtleties with this formula). But we can try to work on a version without the transpose.

[the remark this answers is not in the document]
Comment (rrahir) [reply src/functions/module_text.ts:623]: I don't think it's necessary, it was more a generic comment for the future, but if we are aware of  this, it's fine :)
